### PR TITLE
feat(i18n): add switcher for language

### PR DESCRIPTION
**Pull request by splincode**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows [Conventional Commits](https://www.conventionalcommits.org/en/)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Refactoring
- [ ] Code style update
- [ ] Build or CI related changes
- [ ] Documentation content changes

## What is the current behavior?

Closes # 

## What is the new behavior?

https://user-images.githubusercontent.com/12021443/182381958-12cea9b6-ad1a-4cf9-88c0-27b84fbd49ca.mov

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: projects/addon-doc/src/components/language-switcher/language-switcher.component.html [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        placeholder="Language"
+    />
+
+    <ng-template tuiDataList>
+        <tui-data-list>
+            <button
+                *ngFor="let name of names"
+                tuiOption
+                [value]="name"
+                [textContent]="flags.get(name)"
```
Comment (waterplea) [inline projects/addon-doc/src/components/language-switcher/language-switcher.component.html:19]: This way you overwrite internal structure of `tuiOption` component.

File: projects/addon-doc/src/components/language-switcher/language-switcher.component.html [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+                (click)="set(name)"
+            ></button>
+            <button
+                tuiOption
+                (click)="reset()"
```
Comment (waterplea) [inline projects/addon-doc/src/components/language-switcher/language-switcher.component.html:24]: Not sure we need this option.

File: projects/addon-doc/src/components/language-switcher/language-switcher.component.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+// @dynamic
+@Component({
+    selector: `tui-language-switcher`,
+    templateUrl: `./language-switcher.component.html`,
+    styleUrls: [`./language-switcher.component.less`],
+    changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class TuiLanguageSwitcherComponent {
+    @Input(`tuiTextfieldSize`)
+    size: TuiSizeS | TuiSizeL = `m`;
```
Comment (waterplea) [inline projects/addon-doc/src/components/language-switcher/language-switcher.component.ts:23]: This is unnecessary

File: projects/addon-doc/src/components/language-switcher/language-switcher.component.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+    @Input(`tuiDropdownLimitWidth`)
+    @tuiDefaultProp()
+    limitWidth: TuiDropdownWidthT = `min`;
```
Comment (waterplea) [inline projects/addon-doc/src/components/language-switcher/language-switcher.component.ts:27]: Same here

File: projects/addon-doc/src/components/language-switcher/language-switcher.component.html [hunk 5]
```diff
@@ -0,0 +1,37 @@
+<tui-select
+    [tuiTextfieldSize]="size"
+    [tuiDropdownLimitWidth]="limitWidth"
```
Comment (waterplea) [inline projects/addon-doc/src/components/language-switcher/language-switcher.component.html:3]: Let's remove those

File: projects/addon-doc/src/components/language-switcher/language-switcher.component.ts [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        this.forceUpdateView();
+    }
+
+    reset(): void {
+        this.switcher.removeLang();
+        this.language.setValue(this.switcher.currentLanguage);
+        this.forceUpdateView();
+    }
+
+    forceUpdateView(): void {
```
Comment (waterplea) [inline projects/addon-doc/src/components/language-switcher/language-switcher.component.ts:66]: Why do we need that? Changing the language with select control already marks the view for checking and triggers application tick, since we do that with UI.

File: projects/addon-doc/src/components/language-switcher/language-switcher.module.ts [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+    TuiDataListModule,
+    TuiDropdownControllerModule,
+    TuiTextfieldControllerModule,
+} from '@taiga-ui/core';
+import {TuiDataListWrapperModule, TuiSelectModule} from '@taiga-ui/kit';
+
+import {TuiLanguageSwitcherComponent} from './language-switcher.component';
+
+@NgModule({
+    imports: [
```
Comment (waterplea) [inline projects/addon-doc/src/components/language-switcher/language-switcher.module.ts:14]: Watch out for unused modules.

File: projects/demo/src/modules/customization/i18n/i18n.template.html [hunk 8]
```diff
@@ -3,5 +3,21 @@
     i18n-header
     header="Internationalization"
 >
-    <markdown [data]="(readme | async) || ''"></markdown>
+    <ng-template pageTab="Supported languages">
+        <markdown [data]="(readme | async) || ''"></markdown>
```
Comment (waterplea) [inline projects/demo/src/modules/customization/i18n/i18n.template.html:7]: I wonder if percentages are accurate :D

File: projects/addon-doc/src/components/language-switcher/language-switcher.component.ts [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+    size: TuiSizeS | TuiSizeL = `m`;
+
+    @Input(`tuiDropdownLimitWidth`)
+    @tuiDefaultProp()
+    limitWidth: TuiDropdownWidthT = `min`;
+
+    language = new FormControl(this.switcher.currentLanguage);
+
+    flags: Map<TuiLanguageName, string> = new Map([
+        [`chinese`, `🇨🇳 chinese`],
```
Comment (waterplea) [inline projects/addon-doc/src/components/language-switcher/language-switcher.component.ts:32]: We probably need to capitalize words.

File: projects/i18n/tools/language.ts [hunk 10]
```diff
@@ -1,11 +1,25 @@
-import {InjectionToken} from '@angular/core';
-import {Language} from '@taiga-ui/i18n/interfaces';
+import {inject, InjectionToken} from '@angular/core';
+import {TuiLanguage, TuiLanguageName} from '@taiga-ui/i18n/interfaces';
 import {TUI_ENGLISH_LANGUAGE} from '@taiga-ui/i18n/languages';
 import {Observable, of} from 'rxjs';
 
-export const TUI_LANGUAGE = new InjectionToken<Observable<Language>>(
-    `Language for Taiga UI libraries i18n`,
+export const TUI_DEFAULT_LANGUAGE_NAME = new InjectionToken<TuiLanguageName>(
```
Comment (waterplea) [inline projects/i18n/tools/language.ts:6]: Why do we need all 3?

File: projects/i18n/tools/extract-i18n.ts [hunk 11]
```diff
@@ -1,18 +1,24 @@
...
 // eslint-disable-next-line @typescript-eslint/naming-convention
-export function extractI18n<K extends keyof Language>(
+export function extractI18n<K extends keyof TuiLanguage>(
     key: K,
-): () => Observable<Language[K]> {
-    return () => inject(TUI_LANGUAGE).pipe(map(lang => lang[key]));
+): () => Observable<TuiLanguage[K]> {
+    return () =>
+        inject(TUI_LANGUAGE).pipe(
+            switchMap((streamOrValue: Observable<TuiLanguage> | TuiLanguage) =>
```
Comment (waterplea) [inline projects/i18n/tools/extract-i18n.ts:17]: Why the type change of `TUI_LANGUAGE`?

File: projects/i18n/switch/load-lang.ts [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+
+import {TuiLangChunkLoader} from './options';
+
+export function tuiLoadLanguage(
+    language: TuiLanguageName,
+    loader: Exclude<TuiLangChunkLoader, null | undefined>,
+): Observable<TuiLanguage> {
+    return from(normalizeCommonJSImport(loader(language))).pipe(
+        map((module: any) => module?.[`TUI_${language.toUpperCase()}_LANGUAGE`]),
+    ) as Observable<TuiLanguage>;
```
Comment (waterplea) [inline projects/i18n/switch/load-lang.ts:13]: You should be able to declare generic somewhere here without typecast.

File: projects/i18n/switch/load-lang.ts [hunk 13]
```diff
@@ -0,0 +1,19 @@
+import {TuiLanguage, TuiLanguageName} from '@taiga-ui/i18n/interfaces';
+import {from, Observable} from 'rxjs';
+import {map} from 'rxjs/operators';
+
+import {TuiLangChunkLoader} from './options';
+
+export function tuiLoadLanguage(
+    language: TuiLanguageName,
+    loader: Exclude<TuiLangChunkLoader, null | undefined>,
```
Comment (waterplea) [inline projects/i18n/switch/load-lang.ts:9]: `undefined` is not in the type. Also there's a `NonNullable` utility type in TypeScript ;)

File: projects/i18n/switch/load-lang.ts [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+): Observable<TuiLanguage> {
+    return from(normalizeCommonJSImport(loader(language))).pipe(
+        map((module: any) => module?.[`TUI_${language.toUpperCase()}_LANGUAGE`]),
+    ) as Observable<TuiLanguage>;
+}
+
+async function normalizeCommonJSImport<T>(importPromise: Promise<T>): Promise<T> {
+    // CommonJS `module.exports` is wrapped as `default` in ESModule.
+    return importPromise.then((m: any) => (m.default || m) as T);
```
Comment (waterplea) [inline projects/i18n/switch/load-lang.ts:18]: If `m` is any, do we still need `as T`?

File: projects/i18n/switch/language-chunk-loader-token.ts [hunk 15]
```diff
@@ -0,0 +1,10 @@
+import {InjectionToken} from '@angular/core';
+
+import {TuiLangChunkLoader} from './options';
+
+export const TUI_LANGUAGE_CHUNK_LOADER = new InjectionToken<TuiLangChunkLoader>(
+    `Webpack chunk loader for Taiga UI libraries i18n`,
+    {
+        factory: () => null,
```
Comment (waterplea) [inline projects/i18n/switch/language-chunk-loader-token.ts:8]: Not sure it makes sense to add `null` to `loader` type. Maybe instead you can drop default factory and optionally inject it?

File: projects/addon-doc/src/components/language-switcher/language-switcher.component.html [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+    [valueContent]="content"
+    [formControl]="language"
+>
+    Language
+    <input
+        tuiTextfield
+        placeholder="Language"
```
Comment (waterplea) [inline projects/addon-doc/src/components/language-switcher/language-switcher.component.html:10]: I don't think we need this since there's never no value selected

File: projects/addon-doc/src/components/language-switcher/language-switcher.component.less [hunk 17]
```diff
@@ -0,0 +1,4 @@
+.language-switcher {
```
Comment (waterplea) [inline projects/addon-doc/src/components/language-switcher/language-switcher.component.less:1]: This appears to be not used anywhere

File: projects/i18n/switch/language-switcher-token.ts [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+export function tuiLanguageSwitcher(loader: TuiLangChunkLoader): Provider[] {
+    return [
+        {
+            provide: TUI_LANGUAGE_CHUNK_LOADER,
+            useFactory: () => loader,
+        },
+        {
+            provide: TUI_LANGUAGE,
+            deps: [TuiLanguageSwitcher],
+            useFactory: (switcher: TuiLanguageSwitcher) => switcher.asObservable(),
```
Comment (waterplea) [inline projects/i18n/switch/language-switcher-token.ts:17]: I don't think it's necessary to cast it. We can write `useExisting`

File: projects/i18n/switch/options.ts [hunk 19]
```diff
@@ -0,0 +1,5 @@
+import {TuiLanguageName} from '@taiga-ui/i18n/interfaces';
+
+export const TUI_LANG_STORAGE_KEY = `tuiLang` as const;
```
Comment (waterplea) [inline projects/i18n/switch/options.ts:3]: Should it be token instead maybe?

File: projects/i18n/switch/options.ts [hunk 20]
```diff
@@ -0,0 +1,5 @@
...
+
+export type TuiLangChunkLoader = ((lang: TuiLanguageName) => Promise<unknown>) | null;
```
Comment (waterplea) [inline projects/i18n/switch/options.ts:5]: `TuiHandler<TuiLanguageName, Promise<unknown>`

File: projects/i18n/tools/extract-i18n.ts (hunk 11 above)
Comment (splincode) [reply projects/i18n/tools/extract-i18n.ts:17]: Because we need to set asynchronous data
 
image

File: projects/addon-doc/src/components/language-switcher/language-switcher.component.html (hunk 5 above)
Comment (splincode) [reply projects/addon-doc/src/components/language-switcher/language-switcher.component.html]: fixed

File: projects/addon-doc/src/components/language-switcher/language-switcher.component.html (hunk 16 above)
Comment (splincode) [reply projects/addon-doc/src/components/language-switcher/language-switcher.component.html]: fixed

File: projects/addon-doc/src/components/language-switcher/language-switcher.component.html (hunk 2 above)
Comment (splincode) [reply projects/addon-doc/src/components/language-switcher/language-switcher.component.html]: fixed

File: projects/addon-doc/src/components/language-switcher/language-switcher.component.less (hunk 17 above)
Comment (splincode) [reply projects/addon-doc/src/components/language-switcher/language-switcher.component.less]: fixed

File: projects/addon-doc/src/components/language-switcher/language-switcher.component.ts (hunk 3 above)
Comment (splincode) [reply projects/addon-doc/src/components/language-switcher/language-switcher.component.ts]: fixed

File: projects/addon-doc/src/components/language-switcher/language-switcher.component.ts (hunk 4 above)
Comment (splincode) [reply projects/addon-doc/src/components/language-switcher/language-switcher.component.ts]: fixed

File: projects/addon-doc/src/components/language-switcher/language-switcher.component.ts (hunk 9 above)
Comment (splincode) [reply projects/addon-doc/src/components/language-switcher/language-switcher.component.ts]: fixed

File: projects/addon-doc/src/components/language-switcher/language-switcher.component.ts (hunk 6 above)
Comment (splincode) [reply projects/addon-doc/src/components/language-switcher/language-switcher.component.ts]: fixed

File: projects/addon-doc/src/components/language-switcher/language-switcher.module.ts (hunk 7 above)
Comment (splincode) [reply projects/addon-doc/src/components/language-switcher/language-switcher.module.ts:10]: fixed

File: projects/demo/src/modules/customization/i18n/i18n.template.html (hunk 8 above)
Comment (splincode) [reply projects/demo/src/modules/customization/i18n/i18n.template.html:7]: fixed

File: projects/i18n/switch/language-chunk-loader-token.ts (hunk 15 above)
Comment (splincode) [reply projects/i18n/switch/language-chunk-loader-token.ts]: fixed

File: projects/i18n/switch/language-switcher-token.ts (hunk 18 above)
Comment (splincode) [reply projects/i18n/switch/language-switcher-token.ts]: fixed

File: projects/i18n/switch/load-lang.ts (hunk 13 above)
Comment (splincode) [reply projects/i18n/switch/load-lang.ts]: fixed

File: projects/i18n/switch/load-lang.ts (hunk 12 above)
Comment (splincode) [reply projects/i18n/switch/load-lang.ts]: fixed

File: projects/i18n/switch/load-lang.ts (hunk 14 above)
Comment (splincode) [reply projects/i18n/switch/load-lang.ts]: fixed

File: projects/i18n/switch/options.ts (hunk 19 above)
Comment (splincode) [reply projects/i18n/switch/options.ts]: fixed

File: projects/i18n/switch/options.ts (hunk 20 above)
Comment (splincode) [reply projects/i18n/switch/options.ts]: fixed

File: projects/i18n/tools/language.ts (hunk 10 above)
Comment (splincode) [reply projects/i18n/tools/language.ts]: fixed

File: projects/addon-doc/src/components/language-switcher/language-switcher.component.html (hunk 1 above)
Comment (splincode) [reply projects/addon-doc/src/components/language-switcher/language-switcher.component.html]: fixed

File: projects/addon-doc/src/components/language-switcher/language-switcher.component.html [hunk 21]
```diff
@@ -0,0 +1,26 @@
+<tui-select
+    tuiTextfieldSize="m"
```
Comment (waterplea) [inline projects/addon-doc/src/components/language-switcher/language-switcher.component.html:2]: Please remove this

File: projects/addon-doc/src/components/language-switcher/language-switcher.component.html [hunk 22]
```diff
@@ -0,0 +1,26 @@
...
+            </button>
+        </tui-data-list>
+    </ng-template>
+</tui-select>
+
+<ng-template
+    #content
+    let-name
+>
+    {{ name }}
```
Comment (waterplea) [inline projects/addon-doc/src/components/language-switcher/language-switcher.component.html:25]: This doesn't do anything and can be removed, it is basically `item.toString()` as content, which is the default.

File: projects/addon-doc/src/components/language-switcher/language-switcher.component.html [hunk 23]
```diff
@@ -0,0 +1,26 @@
+<tui-select
+    tuiTextfieldSize="m"
+    [valueContent]="content"
+    [formControl]="language"
+>
+    <ng-content></ng-content>
+    <ng-template tuiDataList>
```
Comment (waterplea) [inline projects/addon-doc/src/components/language-switcher/language-switcher.component.html:7]: We write this as structural directive everywhere, let's keep it consistent.

File: projects/addon-doc/src/components/documentation/documentation.template.html [hunk 24]
```diff
@@ -16,6 +16,26 @@
             {{ texts[3] }}
         </th>
     </tr>
+    <tr class="t-row">
```
Comment (waterplea) [inline projects/addon-doc/src/components/documentation/documentation.template.html:19]: I still think we should move it to a single point, not in the API. Not all components have text prompts so it wouldn't make sense in their API tab, and with text changed from `Language` to `Language of components` I think it works fine as a global select.

File: projects/addon-doc/src/components/language-switcher/language-switcher.component.ts [hunk 25]
```diff
@@ -0,0 +1,32 @@
...
+@Component({
+    selector: `tui-language-switcher`,
+    templateUrl: `./language-switcher.component.html`,
+    changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class TuiLanguageSwitcherComponent {
+    language = new FormControl(this.switcher.currentLanguage);
+
+    flags: Map<TuiLanguageName, string> = new Map([
+        [`Chinese`, `🇨🇳 chinese`],
```
Comment (waterplea) [inline projects/addon-doc/src/components/language-switcher/language-switcher.component.ts:14]: These should be capitalized as well 
image

File: projects/addon-doc/src/components/language-switcher/language-switcher.component.ts [hunk 26]
```diff
@@ -0,0 +1,32 @@
...
+        [`Polish`, `🇵🇱 polish`],
+        [`Portuguese`, `🇵🇹 portuguese`],
+        [`Russian`, `🇷🇺 russian`],
+        [`Spanish`, `🇪🇸 spanish`],
+        [`Turkish`, `🇹🇷 turkish`],
+        [`Ukrainian`, `🇺🇦 ukrainian`],
+        [`Vietnamese`, `🇻🇳 vietnamese`],
+    ]);
+
+    names: TuiLanguageName[] = Array.from(this.flags.keys());
```
Comment (waterplea) [inline projects/addon-doc/src/components/language-switcher/language-switcher.component.ts:29]: nit: all props in this component are `readonly` ;)

File: projects/addon-doc/src/components/language-switcher/language-switcher.module.ts [hunk 27]
```diff
@@ -0,0 +1,22 @@
...
+        CommonModule,
+        FormsModule,
+        TuiSelectModule,
+        TuiDataListModule,
+        ReactiveFormsModule,
+        TuiDataListWrapperModule,
+        TuiTextfieldControllerModule,
```
Comment (waterplea) [inline projects/addon-doc/src/components/language-switcher/language-switcher.module.ts:17]: My last comment still stands. `FormsModule`, `TuiDataListWrapperModule` and `TuiTextfieldControllerModule` are not needed.

File: projects/i18n/switch/language-switcher-token.ts [hunk 28]
```diff
@@ -0,0 +1,19 @@
+import type {Provider} from '@angular/core';
```
Comment (waterplea) [inline projects/i18n/switch/language-switcher-token.ts:1]: I feel like this file should be called `language-switcher.providers.ts` and `TUI_LANGUAGE_CHUNK_LOADER` can be moved here, I don't see a point in splitting these entities into 2 files.

File: projects/i18n/switch/options.ts [hunk 29]
```diff
@@ -0,0 +1,19 @@
...
+
+export const TUI_LANGUAGE_STORAGE_KEY = new InjectionToken<string>(
+    `Default key for search value in storage`,
+    {
+        factory: () => `tuiLang`,
+    },
+);
+
+export const TUI_LANGUAGE_DEFAULT_STORAGE = new InjectionToken<Storage>(
+    `Default key for search value in storage`,
```
Comment (waterplea) [inline projects/i18n/switch/options.ts:15]: Wrong description

File: projects/i18n/switch/options.ts [hunk 30]
```diff
@@ -0,0 +1,19 @@
...
+export type TuiLangChunkLoader = (lang: TuiLanguageName) => Promise<unknown>;
+
+export const TUI_LANGUAGE_STORAGE_KEY = new InjectionToken<string>(
+    `Default key for search value in storage`,
+    {
+        factory: () => `tuiLang`,
+    },
+);
+
+export const TUI_LANGUAGE_DEFAULT_STORAGE = new InjectionToken<Storage>(
```
Comment (waterplea) [inline projects/i18n/switch/options.ts:14]: I was also thinking about abstracting the `Storage` but couldn't think of a reason to. Do you have an example where this might come in handy? Maybe `IonicStorage` but it has different interface — it works with promises. This is probably a bit excessive.

File: projects/i18n/switch/load-lang.ts [hunk 31]
```diff
@@ -0,0 +1,33 @@
...
+
+interface Options {
+    storage: Storage;
+    key: string;
+    fallback: TuiLanguage;
+    loader: TuiLangChunkLoader | null;
+}
+
+export function tuiAsyncLoadLanguage(options: Options): Observable<TuiLanguage> {
+    const {storage, key, fallback, loader} = options;
```
Comment (waterplea) [inline projects/i18n/switch/load-lang.ts:15]: This could be destructed in place.

File: projects/i18n/switch/load-lang.ts [hunk 32]
```diff
@@ -0,0 +1,33 @@
...
+    const language = storage.getItem(key) as TuiLanguageName | null;
```
Comment (waterplea) [inline projects/i18n/switch/load-lang.ts:16]: I think it's a bit too much to pass storage and key in this utility function. Let's just pass `language`.

File: projects/i18n/switch/language-switcher.ts [hunk 33]
```diff
@@ -0,0 +1,53 @@
...
+        private readonly storage: Storage,
+        @Optional()
+        @Inject(TUI_LANGUAGE_CHUNK_LOADER)
+        private readonly loader: TuiLangChunkLoader | null,
+    ) {
+        super(tuiAsyncLoadLanguage({key, storage, fallback, loader}));
+    }
+
+    get currentLanguage(): TuiLanguageName {
+        return (this.storage.getItem(this.key) as TuiLanguageName) ?? this.fallback.name;
```
Comment (waterplea) [inline projects/i18n/switch/language-switcher.ts:32]: Empty string doesn't make sense as `currentLanguage` so let's change `??` to `||` for simplicity and smaller transpiled bundle.

File: projects/i18n/switch/language-chunk-loader-token.ts [hunk 34]
```diff
@@ -0,0 +1,7 @@
+import {InjectionToken} from '@angular/core';
+
+import {TuiLangChunkLoader} from './options';
+
+export const TUI_LANGUAGE_CHUNK_LOADER = new InjectionToken<TuiLangChunkLoader>(
```
Comment (waterplea) [inline projects/i18n/switch/language-chunk-loader-token.ts:5]: Maybe let's drop `CHUNK` for brevity?

File: projects/i18n/switch/options.ts [hunk 35]
```diff
@@ -0,0 +1,19 @@
+import {inject, InjectionToken} from '@angular/core';
+import {LOCAL_STORAGE} from '@ng-web-apis/common';
+import {TuiLanguageName} from '@taiga-ui/i18n/interfaces';
+
+export type TuiLangChunkLoader = (lang: TuiLanguageName) => Promise<unknown>;
```
Comment (waterplea) [inline projects/i18n/switch/options.ts:5]: Let's call it just `TuiLanguageLoader`. Nitpicking, but not a big fan of reducing `Language` to `Lang`. I believe somewhere in our internal codestyle there's an agreement not to do words reducing :)

File: projects/i18n/switch/options.ts [hunk 36]
```diff
@@ -0,0 +1,19 @@
...
+
+export const TUI_LANGUAGE_STORAGE_KEY = new InjectionToken<string>(
+    `Default key for search value in storage`,
+    {
+        factory: () => `tuiLang`,
```
Comment (waterplea) [inline projects/i18n/switch/options.ts:10]: `tuiLanguage`?

File: projects/addon-doc/src/components/language-switcher/language-switcher.component.ts (hunk 25 above)
Comment (nsbarsukov) [reply projects/addon-doc/src/components/language-switcher/language-switcher.component.ts:14]: > These should be capitalized as well image

@splincode oh, according to this screenshot, the flag-smiles do not properly work on Windows OS.

May be we should use the same images as inside `InputPhoneInternational` ?
For example, for English:
https://github.com/Tinkoff/taiga-ui/blob/main/projects/icons/src/GB.png

File: projects/i18n/interfaces/index.ts [hunk 37]
```diff
@@ -1 +1,2 @@
 export * from './language';
+export * from './names';
```
Comment (nsbarsukov) [inline projects/i18n/interfaces/index.ts:2]: nit:
```suggestion
export * from './language-names';
```

File: projects/i18n/switch/language-switcher.ts [hunk 38]
```diff
@@ -0,0 +1,53 @@
...
+        @Inject(TUI_LANGUAGE_DEFAULT_STORAGE)
+        private readonly storage: Storage,
+        @Optional()
+        @Inject(TUI_LANGUAGE_CHUNK_LOADER)
+        private readonly loader: TuiLangChunkLoader | null,
+    ) {
+        super(tuiAsyncLoadLanguage({key, storage, fallback, loader}));
+    }
+
+    get currentLanguage(): TuiLanguageName {
```
Comment (nsbarsukov) [inline projects/i18n/switch/language-switcher.ts:31]: It is `BehaviorSubject`.
Let's rename this getter to `value` to correctly implements [its interface](https://github.com/ReactiveX/rxjs/blob/9a85e01625a1ee7a23e4ab2c2c6d3089a58e2e54/src/internal/BehaviorSubject.ts#L16) ?

File: projects/i18n/switch/language-switcher.ts [hunk 39]
```diff
@@ -0,0 +1,53 @@
...
+        return (this.storage.getItem(this.key) as TuiLanguageName) ?? this.fallback.name;
+    }
+
+    set(language: TuiLanguageName): void {
```
Comment (nsbarsukov) [inline projects/i18n/switch/language-switcher.ts:35]: It is `Subject`.
Let's rename this method to `next` to correctly implements [its interface](https://github.com/ReactiveX/rxjs/blob/9a85e01625a1ee7a23e4ab2c2c6d3089a58e2e54/src/internal/Subject.ts#L60) ?

**Review by nsbarsukov — APPROVED** [review]

File: projects/i18n/switch/language-switcher.ts (hunk 38 above)
Comment (splincode) [reply projects/i18n/switch/language-switcher.ts:31]: image

File: projects/i18n/switch/language-switcher.ts (hunk 39 above)
Comment (splincode) [reply projects/i18n/switch/language-switcher.ts:35]: image

Comment (splincode) [reply projects/i18n/switch/language-switcher.ts:35]: image

File: projects/addon-doc/src/components/language-switcher/language-switcher.component.html (hunk 21 above)
Comment (splincode) [reply projects/addon-doc/src/components/language-switcher/language-switcher.component.html]: fixed

File: projects/addon-doc/src/components/language-switcher/language-switcher.component.html (hunk 22 above)
Comment (splincode) [reply projects/addon-doc/src/components/language-switcher/language-switcher.component.html]: fixed

File: projects/addon-doc/src/components/language-switcher/language-switcher.component.html (hunk 23 above)
Comment (splincode) [reply projects/addon-doc/src/components/language-switcher/language-switcher.component.html]: fixed

File: projects/addon-doc/src/components/documentation/documentation.template.html (hunk 24 above)
Comment (splincode) [reply projects/addon-doc/src/components/documentation/documentation.template.html]: fixed

File: projects/addon-doc/src/components/language-switcher/language-switcher.component.ts (hunk 25 above)
Comment (splincode) [reply projects/addon-doc/src/components/language-switcher/language-switcher.component.ts]: fixed

File: projects/addon-doc/src/components/language-switcher/language-switcher.component.ts (hunk 26 above)
Comment (splincode) [reply projects/addon-doc/src/components/language-switcher/language-switcher.component.ts]: fixed

File: projects/addon-doc/src/components/language-switcher/language-switcher.module.ts (hunk 27 above)
Comment (splincode) [reply projects/addon-doc/src/components/language-switcher/language-switcher.module.ts]: fixed

File: projects/i18n/switch/language-switcher-token.ts (hunk 28 above)
Comment (splincode) [reply projects/i18n/switch/language-switcher-token.ts]: fixed

File: projects/i18n/switch/options.ts (hunk 29 above)
Comment (splincode) [reply projects/i18n/switch/options.ts]: fixed

File: projects/i18n/switch/options.ts (hunk 30 above)
Comment (splincode) [reply projects/i18n/switch/options.ts]: fixed

Comment (splincode) [reply projects/i18n/switch/options.ts]: fixed

File: projects/i18n/switch/load-lang.ts (hunk 31 above)
Comment (splincode) [reply projects/i18n/switch/load-lang.ts]: fixed

File: projects/i18n/switch/load-lang.ts (hunk 32 above)
Comment (splincode) [reply projects/i18n/switch/load-lang.ts]: fixed

File: projects/i18n/switch/language-switcher.ts (hunk 33 above)
Comment (splincode) [reply projects/i18n/switch/language-switcher.ts]: fixed

File: projects/i18n/switch/language-chunk-loader-token.ts (hunk 34 above)
Comment (splincode) [reply projects/i18n/switch/language-chunk-loader-token.ts]: fixed

File: projects/i18n/switch/options.ts (hunk 35 above)
Comment (splincode) [reply projects/i18n/switch/options.ts]: fixed

File: projects/i18n/switch/options.ts (hunk 36 above)
Comment (splincode) [reply projects/i18n/switch/options.ts]: fixed

File: projects/i18n/interfaces/index.ts (hunk 37 above)
Comment (splincode) [reply projects/i18n/interfaces/index.ts]: fixed

**Review by nsbarsukov — APPROVED** [review]

**Comment by splincode** [comment]
@waterplea hello, what do you think about this PR?)

File: projects/i18n/switch/language-switcher.options.ts [hunk 40]
```diff
@@ -0,0 +1,16 @@
+import {InjectionToken} from '@angular/core';
+import {TuiLanguageName} from '@taiga-ui/i18n/interfaces';
+
+export type TuiLanguageLoader = (lang: TuiLanguageName) => Promise<unknown>;
+
+export interface TuiLanguageStorage extends Storage {
```
Comment (waterplea) [inline projects/i18n/switch/language-switcher.options.ts:6]: Shouldn't this be in interfaces folder? Seems unrelated to options.

File: projects/i18n/switch/language-switcher.service.ts [hunk 41]
```diff
@@ -0,0 +1,47 @@
...
+        private readonly storage: TuiLanguageStorage,
+        @Optional()
+        @Inject(TUI_LANGUAGE_LOADER)
+        private readonly loader: TuiLanguageLoader | null,
+    ) {
+        super(tuiAsyncLoadLanguage(storage.getItem(key), loader, fallback));
+    }
+
+    get language(): TuiLanguageName {
+        return (this.storage.getItem(this.key) as TuiLanguageName) || this.fallback.name;
```
Comment (waterplea) [inline projects/i18n/switch/language-switcher.service.ts:33]: Unnecessary typecast now :)

File: projects/i18n/switch/language-switcher.loader.ts [hunk 42]
```diff
@@ -0,0 +1,9 @@
+import {InjectionToken} from '@angular/core';
+
+import {TuiLanguageLoader} from './language-switcher.options';
+
+// @note: cannot be transferred to a shared file
+// ReferenceError: Cannot access 'TUI_LANGUAGE_LOADER' before initialization
+export const TUI_LANGUAGE_LOADER = new InjectionToken<TuiLanguageLoader>(
```
Comment (waterplea) [inline projects/i18n/switch/language-switcher.loader.ts:7]: Maybe we should make `tokens` folder in `i18n` like in the rest of projects and move all tokens there? 🤔 Just for consideration.

**Review by waterplea — APPROVED** [review]

File: projects/i18n/switch/language-switcher.loader.ts (hunk 42 above)
Comment (splincode) [reply projects/i18n/switch/language-switcher.loader.ts]: fixed

File: projects/i18n/switch/language-switcher.options.ts (hunk 40 above)
Comment (splincode) [reply projects/i18n/switch/language-switcher.options.ts]: fixed

File: projects/demo/src/modules/app/app.template.html [hunk 43]
```diff
@@ -23,6 +23,13 @@
                 GitHub
             </a>
         </ng-container>
+        <tui-language-switcher
+            size="m"
```
[the remark this answers is not in the document]
Comment (splincode) [reply projects/demo/src/modules/app/app.template.html]: fixed
